tvm/src/codegen/llvm/codegen_llvm.cc

```diff
@@ -159,7 +159,17 @@ void CodeGenLLVM::AddFunctionInternal(const LoweredFunc& f, bool ret_void) {
     }
   }
   llvm::BasicBlock* entry = llvm::BasicBlock::Create(*ctx_, "entry", function_);
+  llvm::GlobalVariable* assert_flag_global_ = new llvm::GlobalVariable(
+      *module_, llvm::Type::getInt32Ty(*ctx_), false,
+      llvm::GlobalValue::PrivateLinkage, 0, "assert_flag");
```

Comment:
If you are using `new`, you need to free it.

Reply:
Where and how should I free it?

Reply:
https://www.geeksforgeeks.org/new-and-delete-operators-in-cpp-for-dynamic-memory/

Reply:
I'm just not sure where it should be freed because freeing it at the end of AddFunctionInternal causes an llvm error and the variable is not declared outside the scope of that function.

Reply:
You don't need to use a pointer at the very beginning, right? Also, we only add `_` to the end of a variable if it is a class member.

Reply:
Do you mean I should declare it in a different function instead of AddFunctionInternal? I put it there so that I could initialize it to 1 to account for when there aren't any assert statements. Is there a better way to do this?

Reply:
No. First, I believe we don't need to create a pointer. This should be sufficient:
`llvm::GlobalVariable assert_flag_global_ = llvm::GlobalVariable(...)`

Second, this variable is used inside this single function, we don't need to add an `_`. We only add that to a class member variable. In other words, you should just name it `assert_flag_global`.

Reply:
<img width="509" alt="2021-07-20 (2)" src="https://user-images.githubusercontent.com/72524552/126374210-e3f68f16-bfe2-40a6-89c9-b414b14cd3b5.png">

Sorry is this what you had in mind? When I compile it I get an error saying "use of deleted function ‘llvm::GlobalVariable::GlobalVariable(const llvm::GlobalVariable&)’"

Reply:
It seems a copy constructor is prohibited. What about
`llvm::GlobalVariable* assert_flag_global_ = &llvm::GlobalVariable(...)`

Reply:
I guess that wouldn't work also. What's the error when deleting the variable?

Reply:
You can also try to declare it inside `void CodeGenLLVM::Init` and free it inside `std::unique_ptr<llvm::Module> CodeGenLLVM::Finish()`.

Reply:
This is what happens when I delete assert_flag_global at the end of AddFunctionInternal: ` LLVM ERROR: Undefined temporary symbol
`

I noticed that global variables get defined in GetConstString as well. Where do those get freed?

Reply:
Can you try my suggestion first?

Reply:
I got a segfault error when I declared it inside `void CodeGenLLVM::Init` and freed it inside `std::unique_ptr<llvm::Module> CodeGenLLVM::Finish()`

Reply:
Hmmm.... let roll back to your original solution then. Not sure why that's the case but I don't really like using new without delete.